src/mapping/query/function/format_number.rs

```diff
@@ -0,0 +1,254 @@
+}
+
+impl FormatNumberFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        scale: Option<usize>,
+        decimal_separator: Option<&str>,
+        grouping_separator: Option<&str>,
+    ) -> Self {
+        let scale = scale.map(|v| Box::new(Literal::from(Value::from(v as i64))) as _);
+        let decimal_separator =
+            decimal_separator.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+        let grouping_separator =
+            grouping_separator.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+
+        Self {
+            query,
+            scale,
+            grouping_separator,
+            decimal_separator,
+        }
+    }
+}
+
+impl Function for FormatNumberFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let value = required!(ctx, self.query,
+            Value::Integer(v) => Decimal::from_i64(v),
+            Value::Float(v) => Decimal::from_f64(v),
+        )
+        .ok_or("unable to parse number")?;
+
+        let scale = optional!(ctx, self.scale, Value::Integer(v) => v);
+        let grouping_separator = optional!(ctx, self.grouping_separator, Value::Bytes(v) => v);
+        let decimal_separator = optional!(ctx, self.decimal_separator, Value::Bytes(v) => v)
+            .unwrap_or_else(|| Bytes::from("."));
+
+        // Split integral and fractional part of float.
+        let mut parts = value
+            .to_string()
+            .split('.')
+            .map(ToOwned::to_owned)
+            .collect::<Vec<String>>();
+
+        // Manipulate fractional part based on configuration.
+        match scale {
+            Some(i) if i == 0 => parts.truncate(1),
+            Some(i) => {
+                let i = i as usize;
+
+                if parts.len() == 1 {
+                    parts.push("".to_owned())
+                }
+
+                if i > parts[1].len() {
+                    for _ in 0..i - parts[1].len() {
+                        parts[1].push_str("0")
+                    }
+                } else {
+                    parts[1].truncate(i)
+                }
+            }
+            None => {}
+        }
+
+        // Manipulate integral part based on configuration.
+        if let Some(sep) = grouping_separator.as_deref() {
+            let sep = String::from_utf8_lossy(sep);
+            let start = parts[0].len() % 3;
+
+            let positions: Vec<usize> = parts[0]
+                .chars()
+                .skip(start)
+                .enumerate()
+                .map(|(i, _)| i)
+                .filter(|i| i % 3 == 0)
+                .collect();
+
+            for (i, pos) in positions.iter().enumerate() {
+                parts[0].insert_str(pos + (i * sep.len()) + start, &sep);
+            }
+        }
+
+        // Join results, using configured decimal separator.
+        Ok(Value::from(
+            parts.join(&String::from_utf8_lossy(&decimal_separator[..])),
+        ))
```

Comment:
Went through several iterations of this to find something that works, doesn't allocate too much, and is easy to follow, but I'm sure there are many improvements to be made here.

@bruceg anything that jumps out at you?

Reply:
Could you possibly (ab)use `format!` and post-process to add grouping and fix the decimal and grouping characters? Something like `format!("{:.*}", scale, value)`, replace the decimal point, and then possibly add grouping? If not, I would suggest pre-allocating the result string (`String::with_capacity`) and pushing parts into it. There's probably also a way of doing that with an iterator and then `.collect::<String>()`, but how that would work is less obvious to me off hand.

Reply:
Hmm, I just saw an article referencing two crates that may do what you want: [lexical](https://crates.io/crates/lexical) and [ryu](https://crates.io/crates/ryu). Vector already pulls both in from dependencies, though "lexical" looks like the simplest.

Reply:
Unfortunately, neither of those crates did exactly what we want (for example, we don't want to round decimals, but truncate them, and none of them supported adding a grouping separator), and since I already used the `rust_decimal` crate for another remap function, it seemed easiest to use that to go to a string representation, and then manually truncate and add the grouping separator.

> I would suggest pre-allocating the result string (`String::with_capacity`) and pushing parts into it

Thanks, that makes sense 👍

Reply:
Why do we not want to round decimals? Rust, C, Python, and `printf` in the shell all round when a scale is given in a format specifier.

Reply:
Fair point, the issue itself describes the decimals as being truncated. We can change the implementation to round, I was thinking that these functions were already agreed upon in an earlier stage, but I might be wrong there.

I'll add this to my notes, and circle back to it once the other functions have been implemented 👍

Reply:
> Could we open a separate issue to track?

Good point, let me write one up 👍

Reply:
https://github.com/timberio/vector/issues/4411